### ISSUE TITLE
Inconsistency when rollback more than once

### ISSUE THREAD

**Issue by racs4** [Kindelia/Kindelia#63]
When doing the following steps (not a real code):

```
rt.advance(1000)
rt.rollback(900)
let s1 = rt.state()

rt.advance(1000)
rt.rollback(900)
let s2 = rt.state()

rt.advance(1000)
rt.rollback(900)
let s3 = rt.state()
``` 

I get the following results (these prints are in  `fn rollback`):
```
- rolling back from 1000 to 900
- rolled back to 768
- rolling back from 1000 to 900
- rolled back to 0
- rolling back from 1000 to 900
- rolled back to 768
```

There is a test with this code at the recent version of [racs4/rollback-test](https://github.com/racs4/Kindelia/tree/rollback-test).

**Comment by VictorTaelin** [comment]
What happens is the following:

- We arrive on tick 768 for the first time, with rollback=[768,512,0]
- We arrive on tick 1000, with rollback=[992,768,0].
- We go back to tick 768, with rollback=[768,0].
- We arrive at tick 1000, with rollback=[992,976,0].
- We arrive at tick 0, because 768 isn't on rollback.

The root of the issue is that, when we arrivee at 768 for the first time, we
have rollback in a state which is different from we arrive on it for the second
time. That is unavoidable. The question is: how relevant this is in practice? We
could answer this question by emulating a "run" of the blockchain, as follows:

```
fn reorgs(reorg_chance):
  if random() < reorg_chance:
    return 0
  else
    return 1 + reorgs()

fn test_run(blocks, reorg_chance):
  rt = genesis_block()
  loss = 0
  for _ in 0 .. blocks:
    rt.tick()
    tick = rt.get_tick()
    reorg_length = reorgs(reorg_chance)
    rt.rollback(max(tick - reorg_length, 0))
    loss += tick - rt.get_tick()
    rt.advance(tick)
  return loss
```

This is supposed to emulate a real-world node suffering lots of reorgs. Every
time there is a reorg, we count how many blocks we had to recompute. Ideally, we
want loss to be a linear in function of the reorg length.

Of course, it would be better to do a proper analysis, but this can give us a
quick answer.

Also, it seems like, for some reason, we get `run fail` after going back to
block 0. Is this some issue with your test? Or is there a bug when rolling back
to genesis block? Can you see what is going on?

**Comment by VictorTaelin** [comment]
After Rheidner doing the test above, we realized this indeed happened a lot, causing Rollback to lose information and not perform as expected; the expected number of recomputes was higher than assumed under high reorg rates. After examining the issue, I understood the culprit. The solution is patched on the commit above, https://github.com/Kindelia/Kindelia/commit/0338f61a4b0faf5b061f979c90cf54e4855a4d2f. Waiting further tests to confirm this indeed solves the problem. Rollbacks should be consistent, and it should never roll back to 0 again.
